built-in update check and pydfu-based uploader

tehn opened the pull request:
new CLI command `druid update`

- checks `version.txt` from the crow github repo, compares the available version against crow's `^^v` (sets value to 0 if crow is in bootloader mode)
- if newer
  - downloads file (which should just be a raw `crow.dfu` in the crow repo releases)
  - attempts to flash

---

i apologize for the code, i'm not a python person so the code is not idiomatic. i also pull in a few libraries, any optimizations would be welcomed at a later point.

my understanding is that this should work cross-platform, but i haven't tested this on mac or windows.

---

ps. need to update license? or is that inline pydfu header enough?

tehn commented:
extra feature would be a script eraser, so we get rid of those "bonus" things that ship now with the release bundle

trentgill commented:
this is rad!!

will test today. after the CLI version works well we can get it rolled into the TUI version so you can do it all encapsulated. would def be good to get some guidance on how to do it idiomatically, but let's get it *working* as desired first.

csboling commented on src/druid/cli.py:
```diff
@@ -44,6 +49,70 @@ def upload(filename):
+
+      print(">> local version: ", local_version)
+
+      if local_version >= git_data[0]:
```
I think this lexical string comparison may be problematic, for instance:

```python
>>> '1.0.9' >= '1.0.10'
True
```

tehn replied:
fixed. added `packages` dependency

csboling commented on src/druid/cli.py:
```diff
@@ -44,6 +49,70 @@ def upload(filename):
+        os.remove("crow.dfu")
+
+      print("Downloading new version:", git_data[1])
+      wget.download(git_data[1])
```
I think you can download a file with `requests` without needing an additional dependency on `wget`, something like:

```python
res = requests.get(url)
with open('crow.dfu', 'wb') as fwfile:
    fwfile.write(res.content)
```

tehn replied:
works great! removed `wget`

csboling commented on src/druid/cli.py:
```diff
@@ -44,6 +49,70 @@ def upload(filename):
+        os.remove("crow.dfu")
+
+      print("Downloading new version:", git_data[1])
+      wget.download(git_data[1])
```
I think you can download a file with `requests` without needing an additional dependency on `wget`, something like:

```python
res = requests.get(url)
with open('crow.dfu', 'wb') as fwfile:
    fwfile.write(res.content)
```

csboling reviewed (commented):
Will review and test more this evening, these just jumped out from a first scan.

tehn commented:
thank you for these suggestions! exactly what I needed. will follow up with
fixes.

On Fri, Apr 16, 2021, 11:42 AM csboling ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Will review and test more this evening, these just jumped out from a first
> scan.
> ------------------------------
>
> In src/druid/cli.py
> <https://github.com/monome/druid/pull/68#discussion_r614933148>:
>
> > +      except:
> +        print("No crow found, or might be in bootloader mode already...")
> +        local_version = "0"
> +
> +      # crow found: clear script and read version
> +      if local_version != "0":
> +        crow.write("^^c")
> +        time.sleep(1.0)
> +        c = crow.read(1000000)
> +        crow.write("^^v")
> +        tmp = (crow.read(100)).split("'")
> +        local_version = tmp[1][1:]
> +
> +      print(">> local version: ", local_version)
> +
> +      if local_version >= git_data[0]:
>
> I think this lexical string comparison may be problematic, for instance:
>
> >>> '1.0.9' >= '1.0.10'True
>
> ------------------------------
>
> In src/druid/cli.py
> <https://github.com/monome/druid/pull/68#discussion_r614938113>:
>
> > +        crow.write("^^v")
> +        tmp = (crow.read(100)).split("'")
> +        local_version = tmp[1][1:]
> +
> +      print(">> local version: ", local_version)
> +
> +      if local_version >= git_data[0]:
> +        print("Up to date.")
> +        exit()
> +
> +      # delete old crow.dfu if exists
> +      if os.path.exists("crow.dfu"):
> +        os.remove("crow.dfu")
> +
> +      print("Downloading new version:", git_data[1])
> +      wget.download(git_data[1])
>
> I think you can download a file with requests without needing an
> additional dependency on wget, something like:
>
> res = requests.get(url)with open('crow.dfu', 'wb') as fwfile:
>     fwfile.write(res.content)
>
> ------------------------------
>
> In src/druid/cli.py
> <https://github.com/monome/druid/pull/68#discussion_r614938228>:
>
> > +        crow.write("^^v")
> +        tmp = (crow.read(100)).split("'")
> +        local_version = tmp[1][1:]
> +
> +      print(">> local version: ", local_version)
> +
> +      if local_version >= git_data[0]:
> +        print("Up to date.")
> +        exit()
> +
> +      # delete old crow.dfu if exists
> +      if os.path.exists("crow.dfu"):
> +        os.remove("crow.dfu")
> +
> +      print("Downloading new version:", git_data[1])
> +      wget.download(git_data[1])
>
> I think you can download a file with requests without needing an
> additional dependency on wget, something like:
>
> res = requests.get(url)with open('crow.dfu', 'wb') as fwfile:
>     fwfile.write(res.content)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/monome/druid/pull/68#pullrequestreview-637837916>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB4I4EKR2SV6MH3NEKYFU3TJBLERANCNFSM43AIKDUQ>
> .
>

csboling commented on src/druid/pydfu.py:
```diff
@@ -0,0 +1,584 @@
+#!/usr/bin/env python2
```
It might be good to include a github permalink or something to indicate where this file came from and what version it is.

trentgill commented on src/druid/cli.py:
```diff
@@ -44,6 +49,71 @@ def upload(filename):
+
+      # crow found: clear script and read version
+      if local_version != "0":
+        crow.write("^^c")
+        time.sleep(1.0)
```
i think this could just be `crow.write("crow.reset()")` as the goal is just to stop any prints from crow (i think?). the sleep time could be reduced to `0.1` as a result.

i guess if we're making breaking changes in a firmware update it could make sense to do a `^^c`  to avoid a possible immediate error after a successful update. that could definitely result in confusion.

trentgill commented:
i've been testing this a bunch just now, and had a very confusing situation where it would consistently fail at 0% with `Exception: DFU: write memory failed`.

I played around with things trying to understand what the error meant. Made a bunch of changes that didn't help, then finally disabled the `Mass erase` flag and it suddenly got to 45% upload before failure. I ran it again and it succeeded.

Just now I `reset --hard` and it's working as expected. So I really have no idea what the issue was.

//

One issue I am seeing is that every time I run `druid update` it reports:
```
>> git version 2.1.2
>> local version:  2.1.2-4-g5abb605
```
Then downloads & rewrites the firmware (even though they are the same). I can see that this is because the `crow.dfu` file in the 2.1.2 release is actually a few commits past 2.1.2. Seems like an issue that won't be the case for regular releases in the future.

tehn commented:
i saw that DFU error early on when i was trying to use the bin-writing method instead of the .dfu so i'm not sure why it'd have been cropping up for you??

yes, the current `crow.dfu` is a "dummy" file--- basically what i had (the newest). we should put up a correct version prior to merging this.

tehn commented:
changed firmware update command from `druid update` to `druid firmware` (update seemed like druid would update itself, which is maybe not a bad idea--- just an alias to call the relevant pip command?)

also (sorry for combined PR) changed detection to product string instead of VID:PID

trentgill commented:
@tehn 
i just pushed a single change & added a `druid clearscript` command as you mentioned above.

The 1-line change is `pydfu.write_elements` was being sent a `mass_erase==True` which told it *not* to erase the necessary pages. Basically it says "I've already erased all the pages, so you can just write directly", however the pages were not yet erased. This fix stops the `DFU: write memory failed` error.

`druid clearscript` is the same behaviour as the `erase_userscript` bash script / batch file in the crow release zip. It does not attempt to enter bootloader mode, as the only time it should be used is when crow is unresponsive - thus the bootloader must be forced. I believe this will actually be unused with the new firmware as the 3.0 firmware fixes the bug that was causing crow to fail to boot. Still it's nice to have just in case.

tehn commented:
looks great.

final issue is to make sure i didn't introduce a breaking change with the USB device string detection (which i simplified in the newest firmware) so i'll test with both new and old-firmwared crows

trentgill commented:
@tehn 
i was just digging into why this branch has broken the repl, and i think it because of that change. Basically means you can't use the updated druid until you've updated your firmware, which delays the usefulness of this PR until the *next* crow firmware update.

perhaps we can make druid tolerant of old & new style usb identifiers?

tehn commented:
i closed the usb string simplification: it's totally fine as is. i was acting impulsively on a suggestion from elsewhere
